### PR TITLE
Fused MOE for Mixtral

### PR DESCRIPTION
**Pull request by pcmoritz**
This builds on https://github.com/vllm-project/vllm/pull/2453 and https://github.com/vllm-project/vllm/pull/2293 to fuse the MOE kernel for the Mixtral model.

It seems to give a significant performance improvement though (in my setup from 28600 to 33600 tok / s with 1000 input tokens and 50 output tokens on H100).

Latency with `python benchmarks/benchmark_latency.py --model=mistralai/Mixtral-8x7B-Instruct-v0.1 --input-len 1000 --output-len 50 -tp 8 --num-iters 100 --batch-size <bs>`:

This PR:

```
bs=1: 0.459s
bs=2: 0.515s
bs=4: 0.610s
bs=8: 0.813s
bs=16: 1.044s
bs=32: 1.489s
bs=64: 2.419s
```

Compare to master:

```
bs=1: 0.590s
bs=2: 0.631s
bs=4: 0.709s
bs=8: 0.838s
bs=16: 1.086s
bs=32: 1.615s
bs=64: 2.727s
```

### CODE REVIEW

**Comment by pcmoritz** [comment]
MMLU evaluation on this PR looks good as well:

```
|      Groups      |Version|Filter|n-shot|Metric|Value |   |Stderr|
|------------------|-------|------|-----:|------|-----:|---|-----:|
|mmlu              |N/A    |none  |     0|acc   |0.7038|±  |0.1407|
| - humanities     |N/A    |none  |     5|acc   |0.6459|±  |0.1578|
| - other          |N/A    |none  |     5|acc   |0.7763|±  |0.1104|
| - social_sciences|N/A    |none  |     5|acc   |0.8109|±  |0.0704|
| - stem           |N/A    |none  |     5|acc   |0.6143|±  |0.1402|
```

**Comment by pcmoritz** [comment]
Latency numbers:

```
python benchmarks/benchmark_latency.py --model=mistralai/Mixtral-8x7B-Instruct-v0.1 --input-len 1000 --output-len 50 -tp 8 --num-iters 100 --batch-size <bs>
```

This PR:

```
bs=1: 0.561s
bs=2: 0.600s
bs=4: 0.725s
bs=8: 0.912s
bs=16: 1.131s
bs=32: 1.547s
bs=64: 2.43s
```

Master:

```
bs=1: 0.590s
bs=2: 0.631s
bs=4: 0.709s
bs=8: 0.838s
bs=16: 1.086s
bs=32: 1.615s
bs=64: 2.727s
```

So very nice improvements on both throughput and latency (except for some medium batch sizes, but maybe that can be further optimized by tuning the block sizes better).

**Comment by pcmoritz** [comment]
With the latest version of the fused MOE kernel, the fused kernel is now strictly dominating the current master (same settings as above):
```
bs=1: 0.459s
bs=2: 0.515s
bs=4: 0.610s
bs=8: 0.813s
bs=16: 1.044s
bs=32: 1.489s
bs=64: 2.419s
```

**Comment by pcmoritz** [comment]
@WoosukKwon It probably makes sense to review/merge https://github.com/vllm-project/vllm/pull/2453 first since the fused_moe kernel is from there :)

**Comment by casper-hansen** [comment]
@pcmoritz I tried importing your code from here and found that there is a mean maximum difference of `0.3545` in the logits between the normal Mixtral MoE and the fused one.

It seems this is a large difference. Could you add a test between the normal `MixtralSparseMoeBlock` from transformers and the fused module?

**Comment by pcmoritz** [comment]
Thanks @casper-hansen, let me look into this some more / compare the numerics with HuggingFace. Here is what I've figured out so far:

```python
import torch
from transformers import AutoModelForCausalLM

from vllm.model_executor.layers.moe import MoE

from vllm.model_executor.models.mixtral import MixtralModel
from vllm.model_executor.models.mixtral import MixtralForCausalLM


model = AutoModelForCausalLM.from_pretrained("mistralai/Mixtral-8x7B-v0.1")
config = model.config

mixtral_moe = model.model.layers[0].block_sparse_moe

hidden_states = torch.randn((1, 1, 4096))
output = mixtral_moe.forward(hidden_states)
```

vLLM:

```python
vllm_moe = MoE(
    config.num_local_experts,
    config.num_experts_per_tok,
    config.hidden_size,
    config.intermediate_size,
    params_dtype=torch.bfloat16
)

# Load weights:

from vllm.model_executor.weight_utils import hf_model_weights_iterator

expert_params_mapping = [
    ("ws" if weight_name in ["w1", "w3"] else "w2s",
    f"experts.{expert_id}.{weight_name}.weight", expert_id)
    for expert_id in range(config.num_local_experts)
    for weight_name in ["w1", "w2", "w3"]
]

params_dict = dict(vllm_moe.named_parameters())
for name, loaded_weight in hf_model_weights_iterator("mistralai/Mixtral-8x7B-v0.1"):
    if name == "model.layers.0.block_sparse_moe.gate.weight":
        params_dict["gate.weight"] = loaded_weight
    if name.startswith("model.layers.0.block_sparse_moe.experts"):
        for param_name, weight_name, expert_id in expert_params_mapping:
            param = params_dict[param_name]
            weight_loader = param.weight_loader
            weight_loader(param, loaded_weight, weight_name, expert_id=expert_id)
```

**Comment by casper-hansen** [comment]
See the test I created below for reference. I'm not sure what causes the difference, but seems it's a large difference.

https://github.com/casper-hansen/AutoAWQ/blob/mixtral_fused/tests/test_fused_moe.py

File: vllm/model_executor/layers/moe.py
Comment (WoosukKwon) [inline vllm/model_executor/layers/moe.py:1]: @pcmoritz Should we move the `MoE`class back to the Mixtral model file? It seems like this MoE layer is not shared between Mixtral and DeepSeek.

Comment (pcmoritz) [reply vllm/model_executor/layers/moe.py:1]: Sounds good to me! Feel free to make any edits to the PR you'd like to make or let me know if I should make them :)

Comment (WoosukKwon) [reply vllm/model_executor/layers/moe.py:1]: I'd appreciate it if you can do it!

File: vllm/model_executor/layers/moe.py [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+        # router_logits: (batch * sequence_length, n_experts)
+        router_logits, _ = self.gate(hidden_states)
+
+        routing_weights = F.softmax(router_logits, dim=1, dtype=torch.float)
+        routing_weights, selected_experts = torch.topk(routing_weights,
+                                                       self.top_k,
+                                                       dim=-1)
+        routing_weights /= routing_weights.sum(dim=-1, keepdim=True)
+
+        final_hidden_states = self.fused_moe_infer(hidden_states,
```
Comment (WoosukKwon) [inline vllm/model_executor/layers/moe.py:106]: nit:  can we directly use `fused_moe`?

Comment (pcmoritz) [reply vllm/model_executor/layers/moe.py:106]: Yes great suggestion (we did this on the fused MOE PR but hadn't backported it yet)

File: vllm/model_executor/models/mixtral.py [hunk 2]
```diff
@@ -399,14 +299,26 @@ def load_weights(self,
...
+                                  expert_id=expert_id)
+                    break
+                else:
+                    # Skip loading extra bias for GPTQ models.
+                    if name.endswith(".bias") and name not in params_dict:
+                        continue
+                    # Skip experts that are not assigned to this worker.
+                    if ("block_sparse_moe.experts." in name
+                            and name not in params_dict):
+                        continue
```
Comment (WoosukKwon) [inline vllm/model_executor/models/mixtral.py:320]: I believe we don't need this as we now use TP instead of expert parallelism?

Comment (pcmoritz) [reply vllm/model_executor/models/mixtral.py:320]: Yes, you are right!

**Comment by pcmoritz** [comment]
@casper-hansen I don't know if you followed the discussion -- we looked into the numerical differences (https://github.com/vllm-project/vllm/pull/2453#issuecomment-1915992851) and they are due to the TensorFloat tensor cores being used, so it is expected :)

**Review by WoosukKwon — APPROVED** [review]
@pcmoritz LGTM! Thanks for the great work!

**Comment by WoosukKwon** [comment]
BTW, I added @scv119 (who wrote #2293) as a co-author of the PR. Thanks @scv119 for the original PR!

**Comment by pcmoritz** [comment]
Just to be sure, I re-ran MMLU on the latest version of this PR and the result looks good:
```
|      Groups      |Version|Filter|n-shot|Metric|Value |   |Stderr|
|------------------|-------|------|-----:|------|-----:|---|-----:|
|mmlu              |N/A    |none  |     0|acc   |0.7052|±  |0.1374|
| - humanities     |N/A    |none  |     5|acc   |0.6491|±  |0.1522|
| - other          |N/A    |none  |     5|acc   |0.7744|±  |0.1099|
| - social_sciences|N/A    |none  |     5|acc   |0.8109|±  |0.0690|
| - stem           |N/A    |none  |     5|acc   |0.6178|±  |0.1384|
```

**Comment by WoosukKwon** [comment]
Oh BTW, this PR will break the quantization support for Mixtral. 🤦

**Comment by skt7** [comment]
> benchmarks

@pcmoritz are you usinging any specific implememtation to run MMLU benchmark (and others) on LLMs served through vllm, it will be great if you can share the details.
